### PR TITLE
Added object cache for Swift ext storage

### PR DESCRIPTION
**Pull request by PVince81**
This avoid calling the remote API for repeated calls to is_dir,
filetype, etc

I haven't fully tested this yet, but it does seem to be a bit faster.

I'm starting to think we might want to provide a generic cache for remote storages, because the Amazon implementation would also need it. The Dropbox one also has something similar.

@icewind1991

### CODE REVIEW

**Comment by MorrisJobke** [comment]
@PVince81 @DeepDiver1975 I guess this is something for 8.1?

**Comment by PVince81** [comment]
Yes. This only needs review/testing but is a bigger change.
Let's set to 8.1

**Comment by PVince81** [comment]
One alternative would be to do it in a generic way for all storages by implementing a stat cache as storage wrapper: https://github.com/owncloud/core/issues/13971

**Comment by PVince81** [comment]
Ok, turns out the storage wrapper will not help for local function calls because the storage only calls itself, not the wrapper.

So this means that this PR is still valid (and also that a similar approach like this PR is needed for other storages)

**Comment by LukasReschke** [comment]
Rebase required?

**Comment by PVince81** [comment]
Rebased. Fortunately only a namespace change.

**Comment by enoch85** [comment]
@PVince81 Sorry, not using External Storage anymore. Would have tested otherwise.

**Comment by PVince81** [comment]
@butonic @icewind1991 any chance for a review ?

File: apps/files_external/lib/swift.php [hunk 1]
```diff
@@ -145,6 +178,7 @@ public function mkdir($path) {
 			$metadataHeaders = DataObject::stockHeaders(array());
 			$allHeaders = $customHeaders + $metadataHeaders;
 			$this->getContainer()->uploadObject($path, '', $allHeaders);
```
Comment (icewind1991) [inline apps/files_external/lib/swift.php]: `uploadObject()` returns a `DataObject` which you can put in the cache

File: apps/files_external/lib/swift.php [hunk 2]
```diff
@@ -381,8 +427,11 @@ public function touch($path, $mtime = null) {
 				$path .= '/';
 			}
 
-			$object = $this->getContainer()->getPartialObject($path);
-			$object->saveMetadata($metadata);
+			$object = $this->fetchObject($path);
+			if ($object->saveMetadata($metadata)) {
+				// invalidate target object to force repopulation on fetch
+				$this->objectCache->remove($path);
```
Comment (icewind1991) [inline apps/files_external/lib/swift.php]: is it possible to change the meta data of our existing object so we dont need to get the meta data for this file again

File: apps/files_external/lib/swift.php [hunk 3]
```diff
@@ -405,8 +456,10 @@ public function copy($path1, $path2) {
 			$this->unlink($path2);
 
 			try {
-				$source = $this->getContainer()->getPartialObject($path1);
+				$source = $this->fetchObject($path1);
 				$source->copy($this->bucket . '/' . $path2);
+				// invalidate target object to force repopulation on fetch
+				$this->objectCache->remove($path2);
```
Comment (icewind1991) [inline apps/files_external/lib/swift.php]: set the object from `$path1` with it's path changed?

File: lib/private/cache/arraycache.php [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+ * See the COPYING-README file.
+ */
+
+namespace OC\Cache;
+
+/**
+* Implementation of a volatile cache backed by memory (in form of a simple PHP
+* array) for special purposes.
+*/
+class ArrayCache implements \OCP\ICache {
```
Comment (icewind1991) [inline lib/private/cache/arraycache.php]: https://github.com/owncloud/core/blob/master/lib/private/memcache/arraycache.php ?

Comment (PVince81) [inline lib/private/cache/arraycache.php]: @bantu ^ (the new arraycache impl was your idea)

**Comment by PVince81** [comment]
- [ ] use the existing ArrayCache instead
- [ ] Fix @icewind1991's comments

File: apps/files_external/lib/swift.php (hunk 1 above)
Comment (PVince81) [inline apps/files_external/lib/swift.php]: I prefer to invalidate it. From my experience the cached object might not contain all final attributes which can only be retrieved the next time. (but I might be mixing this up with S3)

**Comment by PVince81** [comment]
@icewind1991 actually you might be right with the ability to cache objects after upload and co without rechecking, but will need deeper digging into the API.

Let me rebase this first.

**Comment by DeepDiver1975** [comment]
hmmm .. in case we move over to flysystem we can make use of their caching mechanisms - I wonder if we should invest time into this right now.

**Comment by DeepDiver1975** [comment]
@PVince81 9.0 ???

**Comment by PVince81** [comment]
Ok, I can revive this for 9.0.
I'm not even sure whether flysystem have a cache mechanism anyway...

**Comment by Xenopathic** [comment]
It seems that Flysystem performs no caching at all...

**Comment by DeepDiver1975** [comment]
> It seems that Flysystem performs no caching at all...

aha .... http://flysystem.thephpleague.com/caching/

**Comment by Xenopathic** [comment]
@DeepDiver1975 Or that :see_no_evil:

**Comment by PVince81** [comment]
Rebased.

Has failing tests:
```
2) Test\Files\Storage\Swift::testRenameDirectory
Failed asserting that false is true.

/var/www/html/owncloud/tests/lib/files/storage/storage.php:468

3) Test\Files\Storage\Swift::testCopyDirectory
Failed asserting that false is true.

/var/www/html/owncloud/tests/lib/files/storage/storage.php:522
```

**Comment by PVince81** [comment]
Rebased and fixed unit tests.
Also now using the CappedCache for the stat cache.

Please review @owncloud/filesystem 
@owncloud/qa can you help regression test this ? All file operations should still work, this makes SWIFT external storage much faster.

**Comment by PVince81** [comment]
also @tjdett @MorrisJobke in case you are interested

**Comment by icewind1991** [comment]
:+1:

**Comment by PVince81** [comment]
Rebased due to the PSR-4 changes. (no conflicts, automatically resolved)

**Comment by PVince81** [comment]
Ran the Swift tests locally and they passed, merging.

**Pull request merged**